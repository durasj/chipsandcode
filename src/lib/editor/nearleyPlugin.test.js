import { describe, expect, it } from 'vitest';
import contentPlugin from './nearleyPlugin';

describe('Nearley Vite Plugin', () => {
  it('Has correct structure', () => {
    const plugin = contentPlugin();

    expect(plugin).toMatchObject({
      name: 'vite-plugin-nearley',
      enforce: 'pre',
      transform: expect.any(Function),
    });
  });

  it('Ignores files that are not .ne', () => {
    const plugin = contentPlugin();

    expect(plugin.transform('const x = 123;', 'component.ts')).toBeUndefined();
    expect(plugin.transform('Some Content', 'component.svelte')).toBeUndefined();
    expect(plugin.transform('* { color: #000000 !important; }', 'component.css')).toBeUndefined();
    expect(plugin.transform('# Something', 'content.md')).toBeUndefined();
  });

  it('Processes grammar into code', () => {
    const plugin = contentPlugin();

    const grammar = `
P -> S

S -> S "+" M
    | M

M -> M "*" T
    | T

T -> "1"
    | "2"
    | "3"
    | "4"
    `;

    expect(plugin.transform(grammar, 'grammar.ne')).toMatchSnapshot();
  });
});
