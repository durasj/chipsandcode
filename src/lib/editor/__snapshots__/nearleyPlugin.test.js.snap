// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nearley Vite Plugin > Processes grammar into code 1`] = `
{
  "code": "// Generated automatically by nearley, version unknown
// http://github.com/Hardmath123/nearley
function id(x) { return x[0]; }
let Lexer = undefined;
let ParserRules = [
    {\\"name\\": \\"P\\", \\"symbols\\": [\\"S\\"]},
    {\\"name\\": \\"S\\", \\"symbols\\": [\\"S\\", {\\"literal\\":\\"+\\"}, \\"M\\"]},
    {\\"name\\": \\"S\\", \\"symbols\\": [\\"M\\"]},
    {\\"name\\": \\"M\\", \\"symbols\\": [\\"M\\", {\\"literal\\":\\"*\\"}, \\"T\\"]},
    {\\"name\\": \\"M\\", \\"symbols\\": [\\"T\\"]},
    {\\"name\\": \\"T\\", \\"symbols\\": [{\\"literal\\":\\"1\\"}]},
    {\\"name\\": \\"T\\", \\"symbols\\": [{\\"literal\\":\\"2\\"}]},
    {\\"name\\": \\"T\\", \\"symbols\\": [{\\"literal\\":\\"3\\"}]},
    {\\"name\\": \\"T\\", \\"symbols\\": [{\\"literal\\":\\"4\\"}]}
];
let ParserStart = \\"P\\";
export default { Lexer, ParserRules, ParserStart };
",
  "map": null,
}
`;
